src/iter/as_parallel.rs

```diff
@@ -0,0 +1,161 @@
+    {
+        let split_count = AtomicUsize::new(current_num_threads());
+        let deque = Deque::new();
+        let stealer = deque.stealer();
+        let done = AtomicBool::new(false);
+        let iter = Mutex::new((self.iter, deque));
+
+        bridge_unindexed(IterParallelProducer {
+            split_count: &split_count,
+            done: &done,
+            iter: &iter,
+            items: stealer,
+        }, consumer)
+    }
+}
+
+struct IterParallelProducer<'a, Iter: Iterator + 'a> {
+    split_count: &'a AtomicUsize,
+    done: &'a AtomicBool,
+    iter: &'a Mutex<(Iter, Deque<Iter::Item>)>,
+    items: Stealer<Iter::Item>,
+}
+
+// manual clone because T doesn't need to be Clone, but the derive assumes it should be
+impl<'a, Iter: Iterator + 'a> Clone for IterParallelProducer<'a, Iter> {
+    fn clone(&self) -> Self {
+        IterParallelProducer {
+            split_count: self.split_count,
+            done: self.done,
+            iter: self.iter,
+            items: self.items.clone(),
+        }
+    }
+}
+
+impl<'a, Iter: Iterator + Send + 'a> UnindexedProducer for IterParallelProducer<'a, Iter>
+    where Iter::Item: Send
+{
+    type Item = Iter::Item;
+
+    fn split(self) -> (Self, Option<Self>) {
+        let mut count = self.split_count.load(Ordering::SeqCst);
+
+        loop {
+            let done = self.done.load(Ordering::SeqCst);
+            match count.checked_sub(1) {
+                Some(new_count) if !done => {
+                    let last_count = self.split_count.compare_and_swap(count, new_count, Ordering::SeqCst);
+                    if last_count == count {
+                        return (self.clone(), Some(self));
+                    } else {
+                        count = last_count;
+                    }
+                }
+                _ => {
+                    return (self, None);
+                }
+            }
+        }
+    }
+
+    fn fold_with<F>(self, mut folder: F) -> F
+        where F: Folder<Self::Item>
+    {
+        loop {
+            match self.items.steal() {
+                Steal::Data(it) => {
+                    folder = folder.consume(it);
+                    if folder.full() {
+                        return folder;
+                    }
+                }
+                Steal::Empty => {
+                    if self.done.load(Ordering::SeqCst) {
+                        // the iterator is out of items, no use in continuing
+                        return folder;
+                    } else {
+                        // our cache is out of items, time to load more from the iterator
+                        match self.iter.try_lock() {
+                            Ok(mut guard) => {
+                                let count = current_num_threads();
+                                let count = (count * count) * 2;
+
+                                let (ref mut iter, ref deque) = *guard;
+
+                                for _ in 0..count {
```

Comment:
Instead of reading exactly `count` items, how about `while deque.len() < count`?  This way, I'm imagining that one thread could become the de facto reader thread, just pushing items as fast as the others can pop them out.  But if the other threads are too busy, this thread will fill the deque up to `count` and then break out to resume processing items itself.  This will be dynamic too, so when other threads get the time they'll start cooperating again.